BlueprintUI/Sources/Internal/CacheTree.swift

```diff
@@ -0,0 +1,52 @@
+import CoreGraphics
+import os.log
+
+/// A size cache that also holds subcaches.
+protocol CacheTree: AnyObject {
+
+    /// The name of this cache
+    var name: String { get }
+
+    /// A reference to use for logging
+    var signpostRef: AnyObject { get }
+
+    /// The sizes that are contained in this cache, keyed by size constraint.
+    subscript(constraint: SizeConstraint) -> CGSize? { get set }
+
+    /// Gets a subcache identified by the given key, or creates a new one.
+    func subcache(key: SubcacheKey, name: @autoclosure () -> String) -> CacheTree
+}
+
+struct SubcacheKey: RawRepresentable, Hashable {
+    /// A key indicating that this will be the only subcache
+    static let singleton = SubcacheKey(rawValue: -1)
+
+    let rawValue: Int
+}
+
+extension CacheTree {
+    /// Convenience method to get a cached size, or compute and store one if it is not in the cache.
+    func get(_ constraint: SizeConstraint, orStore calculation: (SizeConstraint) -> CGSize) -> CGSize {
+        if let size = self[constraint] {
+            return size
+        }
+        let size = calculation(constraint)
+        self[constraint] = size
+        return size
+    }
+
+    /// Gets a subcache for an element with siblings.
+    func subcache(index: Int, element: Element) -> CacheTree {
+        subcache(key: SubcacheKey(rawValue: index), element: element)
+    }
+
+    /// Gets a subcache for an element.
+    func subcache(key: SubcacheKey = .singleton, element: Element) -> CacheTree {
+        subcache(
+            key: key,
+            name: key == .singleton
+                ? "\(self.name).\(type(of: element))"
+                : "\(self.name)[\(key.rawValue)].\(type(of: element))"
```

Comment:
Curious what sort of cost these strings getting built have. Wonder if we could make them memoized to avoid needing to do it during every traversal.

Reply:
> Curious what sort of cost these strings getting built have. Wonder if we could make them memoized to avoid needing to do it during every traversal.

Yeah, I had the same concern, and that's the reason for the autoclosure. It's only hit once per element per traversal, so I think it's pretty small compared to the amount of other work we do during layout. Dunno if we can do much better.

Reply:
Ahh missed the autoclosure – and it's only used for signpost logging, meaning we won't hit it in release, right?

Reply:
> Ahh missed the autoclosure – and it's only used for signpost logging, meaning we won't hit it in release, right?

It _will_ be hit in release, since profiling is typically done against a release build.

Reply:
Thought / ask: Worth adding a way to easily short circuit the signpost logging if we want to do a pure perf trace without the signpost logging appearing? When I was testing, this was always visible in the traces, which while easy to ignore at a top level if inverting the call tree, using something like top functions means it can be a bit harder to realize that the stuff you're trying to optimize is actually just from the signpost logging.

Also, any idea if this stuff fires in device beta or app store builds? When I was testing w/ this in the checkout applet, of the ~15s of so of CPU time, the signpost logging took ~600ms, so its definitely not free.

Reply:
> Thought / ask: Worth adding a way to easily short circuit the signpost logging if we want to do a pure perf trace without the signpost logging appearing? When I was testing, this was always visible in the traces, which while easy to ignore at a top level if inverting the call tree, using something like top functions means it can be a bit harder to realize that the stuff you're trying to optimize is actually just from the signpost logging.

Sure. The [WWDC video](https://developer.apple.com/videos/play/wwdc2018/405/?time=863) on this talks about how they're meant to be super lightweight, but also gives a nice way to disable them.

> Also, any idea if this stuff fires in device beta or app store builds? When I was testing w/ this in the checkout applet, of the ~15s of so of CPU time, the signpost logging took ~600ms, so its definitely not free.

An app store build is still just a release build, so I would guess yes? But I would also guess that in a release build it's much faster than that.

Reply:
Config added! This is now off by default. To enable it, there is a global config at `BlueprintLogging.enabled`.